### PR TITLE
Add attributes advice API (just `DoubleCounter` for now)

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
Implements https://github.com/open-telemetry/opentelemetry-specification/pull/3546

I only added implementation for the `DoubleCounter` in this PR; I'll extend it to all other instruments in the following PR if this one is accepted.

### CODE REVIEW

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/ViewRegistry.java [hunk 1]
```diff
@@ -246,4 +254,18 @@ private static Pattern toRegexPattern(String globPattern) {
     }
     return Pattern.compile(patternBuilder.toString());
   }
+
+  private static boolean shouldApplyAdvice(Advice descriptor) {
+    return descriptor.getAttributes() != null;
```
Comment (breedx-splk) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/ViewRegistry.java:259]: Consider adding `boolean Advice.hasAttributes()` to tidy this up a bit.

File: sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/internal/view/AdviceAttributesProcessorTest.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        processor.process(
+            Attributes.builder()
+                .put(stringKey("abc"), "abc")
+                .put(stringKey("def"), "def")
+                .put(longKey("ghi"), 42L)
+                .put(stringKey("xyz"), "xyz")
+                .build(),
+            Context.root());
+
+    // does not contain key "ghi" because the type is different (stringKey != longKey)
```
Comment (breedx-splk) [inline sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/internal/view/AdviceAttributesProcessorTest.java:41]: 👍🏻 good call to cover this!

**Review by breedx-splk — APPROVED** [review]
Looks good to me.

File: sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/AttributesAdviceTest.java [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+    assertThat(reader.collectAllMetrics())
+        .satisfiesExactly(
+            metric ->
+                assertThat(metric)
+                    .hasDoubleSumSatisfying(
+                        sum -> sum.hasPointsSatisfying(point -> point.hasAttributes(ATTRIBUTES))));
+  }
+
+  @Test
+  void counterWithAdvice() {
```
Comment (jack-berg) [inline sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/AttributesAdviceTest.java:55]: Hey so I made a [comment](https://github.com/open-telemetry/opentelemetry-specification/issues/3622#issuecomment-1654613883) over in the spec which states:

> Just so we're clear, my interpretation is that because advice is non-identifying, two histograms with the same identity and different advices will result in 1 metric stream. And thus we have to pick one set of advice to apply to the resulting metric stream, which should naturally be the first we see.

But that isn't the case as implemented today (and also not for the histogram bucket advice). You can demonstrate this by creating another counter with the same name, and with no advice, and seeing that there are two metrics produced instead of just one:
```
    DoubleCounter counter2 = meterProvider.get("meter").counterBuilder("counter").ofDoubles().build();
    counter2.add(1, ATTRIBUTES);
```

The issue stems from [MetricStorageRegistry#register(MetricStorage)](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/state/MetricStorageRegistry.java#L57-L87), a tricky little piece of code responsible for evaluating whether a newly registered instrument should get a new storage (and thus a new metric stream at collection) or should send its measurements to an existing storage (and thus merge with an existing metric stream at collection).

The problem is that we use `MetricDescriptor.equals(..)` to determine if an existing storage matches, but that is a pretty strict equality, checking equality of name, description, view, InstrumentDescriptor (including name, description, unit, type, value type, and advice), and aggregation name. 

And I believe there's a bug where if a instrument matches except the case of the name, we produce two metric streams, instead of one like I thought we did. See [spec issue #3626](https://github.com/open-telemetry/opentelemetry-specification/pull/3626). 

I think this is probably fine for this PR. I need to go back and fix `MetricStorageRegistry#register` in a more general capacity.

**Review by jack-berg — COMMENTED** [review]
Looks good, just one small edge case to think about in ViewRegistry

File: sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/AttributesAdviceTest.java (hunk 3 above)
Comment (jack-berg) [reply sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/AttributesAdviceTest.java:55]: Opened #5683 to discuss.

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/ViewRegistry.java [hunk 4]
```diff
@@ -127,7 +128,14 @@ public List<RegisteredView> findViews(
 
     // No views matched, use default view
     RegisteredView instrumentDefaultView =
-        Objects.requireNonNull(instrumentDefaultRegisteredView.get(descriptor.getType()));
+        requireNonNull(instrumentDefaultRegisteredView.get(descriptor.getType()));
+
+    // if the user defined an advice, use it
+    if (shouldApplyAdvice(descriptor.getAdvice())) {
+      instrumentDefaultView =
+          applyAdviceToDefaultView(instrumentDefaultView, descriptor.getAdvice());
```
[the remark this answers is not in the document]
Comment (mateuszrzeszutek) [reply sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/ViewRegistry.java:136]: Good point 👍 
Thanks, I'll fix that.

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/AdviceAttributesProcessor.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    return builder.build();
+  }
+
+  @Override
+  public boolean usesContext() {
+    return false;
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
[the remark this answers is not in the document]
Comment (mateuszrzeszutek) [reply sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/AdviceAttributesProcessor.java:38]: I added it specifically for tests, to compare `RegisteredView`s. E.g.

```java
        .isEqualTo(
            Collections.singletonList(
                RegisteredView.create(
                    DEFAULT_REGISTERED_VIEW.getInstrumentSelector(),
                    DEFAULT_VIEW,
                    new AdviceAttributesProcessor(
                        Arrays.asList(stringKey("key1"), stringKey("key2"))),
                    MetricStorage.DEFAULT_MAX_CARDINALITY,
                    SourceInfo.noSourceInfo())));
```

Should I use something else to assert on that? Reflection?

File: sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/AttributesAdviceTest.java (hunk 3 above)
Comment (mateuszrzeszutek) [reply sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/AttributesAdviceTest.java:55]: Yeah, I 100% agree advice should be non-identifying. 

> I think this is probably fine for this PR. I need to go back and fix `MetricStorageRegistry#register` in a more general capacity.

👍

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/AdviceAttributesProcessor.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jack-berg) [reply sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/AdviceAttributesProcessor.java:38]: I think in the past I've done something like:
```
.isInstanceOfSatisfying(AdviceAttributesProcessor.class, processor -> {
  assertThat(processor.getAttributeKeys()).isEqualTo(Arrays.asList(..))
})
```
This just avoids having to maintain an equals / hashcode implementation for something that really isn't normally used as a key in a map, or checked for equality for something else.

**Review by jack-berg — APPROVED** [review]

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/AdviceAttributesProcessor.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (mateuszrzeszutek) [reply sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/internal/view/AdviceAttributesProcessor.java]: Done 👍 
I replaced that with some AssertJ magic, PTAL

File: sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/AttributesAdviceTest.java (hunk 3 above)
Comment (jack-berg) [reply sdk/metrics/src/test/java/io/opentelemetry/sdk/metrics/AttributesAdviceTest.java:55]: Solved in #5701.
